docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,335 @@
+copy, the enclave frees the content of `b`. See the examples of ECALL and OCALL design as
+follows.
+
+- ECALL
+  - Host
+
+    The host invokes the oeedger8r-generated ECALL stub as usual.
+    ```
+    // Oeedger8r-generated code
+    void foo(oe_enclave* enclave, Blob *b)
+    {
+      ...
+      oe_call_enclave_function(...);
+      ...
+    }
+    ```
+
+  - Enclave
+
+    Upon the dispatching of the ECALL, the oeedger8r-generated
+    `ecall_foo` is invoked as usual. However, right after the
+    invocation of the user-defined `foo`, the design adds a piece
+    of code that allocates host memory and copies the content
+    of `b` to the allocated memory.
+
+    ```
+    // Oeedger8r-generated code
+    void ecall_foo()
+    {
+      ...
+      foo(_pargs_in->b);
+      ...
+      /* _pargs_out holds the pointer of b on the host. */
+      if (_pargs_out->b)
+      {
+        Blob *_rhs = (Blob *) _pargs_in->b;
+        _pargs_out->len = _rhs->len;
+        _pargs_out->b = NULL;
+        if (_rhs->b)
+        {
+          /* Use a local variable to prevent the host from
+             manipulating the address value used by memcpy below. */
+          char* tmp_buf = oe_host_malloc(_rhs->len);
+          _pargs_out->buf = tmp_buf;
+          if (!tmp_buf)
+          {
+            _result = OE_FAILURE;
+            goto done;
+          }
+          memcpy(tmp_buf, _rhs->buf, _rhs->len);
+        }
+      }
+      ...
+    done:
+      /* Free host buffers on failure. */
+      if (_result != OE_OK)
+      {
+        if (_pargs_out->b)
+        {
+          oe_host_free(_pargs_out->b->buf);
+          _pargs_out->b->buf = NULL;
+        }
+      }
+      /* Always free the intermediate enclave buffers. */
+      if (_pargs_in->b)
+      {
+        oe_free(_pargs_in->b->buf);
+        _pargs_in->b->buf = NULL;
+      }
+    }
+    ```
+
+- OCALL
+  - Enclave
+
+    Given that the shared memory is available (i.e., the enclave
+    can access host memory), the design of the OCALL adds the similar
+    piece of the code on the enclave side that allocates the memory
+    and copies the content in the host memory to the allocated memory.
+
+    ```
+    // Oeedger8r-generated code
+    void foo(Blob *b)
+    {
+      ...
+      oe_call_host_function(...);
+      ...
+      /* Initialize the deep-copyable, out-only parameters pointed by _pargs_out to NULL.
+         These parameters will be freed at the end of the function if not NULL. */
+      pargs_out->b = NULL;
+
+      if (b)
+      {
+        _pargs_out->b = (Blob *) (_output_buffer + _output_buffer_offset);
+        OE_ADD_SIZE(_output_buffer_offset, size(Blob));
+        b->len = _pargs_out->b->len;
+        b->buf = NULL;
+        if (_pargs_out->b->buf)
+        {
+          b->buf = oe_malloc(b->len);
+          if (!b->buf)
+          {
+            _result = OE_FAILURE;
+            goto done;
+          }
+          memcpy(b->buf, _pargs_out->b->buf, b->len);
```

Comment:
If higher level code did not verify that `_pargs_out->b->buf`, with length of `b->len`, is fully outside of the enclave memory already, the code here must verify that before performing the copy.

Reply:
`b->len` is inside the enclave while `_pargs_out->b->len` is on the hosot.

Reply:
The design has been modified and hence the comments are no longer applicable.